File: src/test/java/net/slipp/support/web/argumentresolver/LoginUserHandlerMethodArgumentResolverTest.java
package net.slipp.support.web.argumentresolver;

import static org.hamcrest.CoreMatchers.*;
import static org.junit.Assert.*;
import static org.mockito.Mockito.*;

import java.lang.annotation.Annotation;

import net.slipp.domain.user.SocialUser;
import net.slipp.support.security.LoginRequiredException;
import net.slipp.support.security.SessionService;

import org.junit.Before;
import org.junit.Test;
import org.junit.runner.RunWith;
import org.mockito.Mock;
import org.mockito.runners.MockitoJUnitRunner;
import org.springframework.core.MethodParameter;
import org.springframework.web.context.request.NativeWebRequest;

@RunWith(MockitoJUnitRunner.class)
public class LoginUserHandlerMethodArgumentResolverTest {
    @Mock
    private MethodParameter parameter;

    @Mock
    private NativeWebRequest webRequest;

    @Mock
    private SessionService sessionService;

    @Mock
    private SocialUser user;

    private LoginUserHandlerMethodArgumentResolver loginUserHandlerMethodArgumentResolver;

    @Before
    public void setUp() throws Exception {
        loginUserHandlerMethodArgumentResolver = new LoginUserHandlerMethodArgumentResolver();
        loginUserHandlerMethodArgumentResolver.setSessionService(sessionService);

        when(user.isGuest()).thenReturn(false);
    }

    @Test
    public void supportsParameter_false() {
        when(parameter.hasParameterAnnotation(LoginUser.class)).thenReturn(false);

        assertThat(loginUserHandlerMethodArgumentResolver.supportsParameter(parameter), is(false));
    }

    @Test
    public void supportsParameter_true() {
        when(parameter.hasParameterAnnotation(LoginUser.class)).thenReturn(true);

        assertThat(loginUserHandlerMethodArgumentResolver.supportsParameter(parameter), is(true));
    }

    @Test(expected = LoginRequiredException.class)
    public void loginUserRequired_but_guest() throws Exception {
        when(parameter.getParameterAnnotation(LoginUser.class)).thenReturn( new FakeLoginUser(true));
        when(sessionService.getLoginUser()).thenReturn(SocialUser.GUEST_USER);

        loginUserHandlerMethodArgumentResolver.resolveArgument(parameter, null, webRequest, null);
    }

    @Test
    public void loginUser_not_required() throws Exception {
        when(parameter.getParameterAnnotation(LoginUser.class)).thenReturn( new FakeLoginUser(false));
        when(sessionService.getLoginUser()).thenReturn(SocialUser.GUEST_USER);

        SocialUser loginUser = (SocialUser) loginUserHandlerMethodArgumentResolver.resolveArgument(parameter, null,
                webRequest, null);

        assertThat(loginUser, is(SocialUser.GUEST_USER));
    }

    @Test
    public void loginUser_일반상황() throws Exception {
        SocialUser user = new SocialUser();
        when(parameter.getParameterAnnotation(LoginUser.class)).thenReturn( new FakeLoginUser(true));
        when(sessionService.getLoginUser()).thenReturn(user);

        SocialUser loginUser = (SocialUser) loginUserHandlerMethodArgumentResolver.resolveArgument(parameter, null,
                webRequest, null);

        assertThat(loginUser, is(user));
    }
    
    public static class FakeLoginUser implements LoginUser {
        private boolean requiredValue = true;

        public FakeLoginUser(boolean requiredValue) {
            this.requiredValue = requiredValue;
        }

        @Override
        public Class<? extends Annotation> annotationType() {
            return LoginUser.class;
        }

        @Override
        public boolean required() {
            return requiredValue;
        }
    }
}

File: src/test/scala/net/slipp/support/web/argumentresolver/LoginUserHandlerMethodArgumentResolverTest.scala
package net.slipp.support.web.argumentresolver

import net.slipp.domain.user.SocialUser
import net.slipp.support.security.{LoginRequiredException, SessionService}
import org.hamcrest.CoreMatchers.is
import org.junit.Assert.assertThat
import org.junit.{Before, Test}
import org.junit.runner.RunWith
import org.mockito.Mock
import org.mockito.Mockito.when
import org.mockito.runners.MockitoJUnitRunner
import org.springframework.core.MethodParameter
import org.springframework.web.context.request.NativeWebRequest

@RunWith(classOf[MockitoJUnitRunner])
class LoginUserHandlerMethodArgumentResolverTest {
  @Mock private var parameter: MethodParameter = null
  @Mock private var webRequest: NativeWebRequest = null
  @Mock private var sessionService: SessionService = null
  @Mock private var user: SocialUser = null
  private var loginUserHandlerMethodArgumentResolver: LoginUserHandlerMethodArgumentResolver = null

  @Before
  def setUp {
    loginUserHandlerMethodArgumentResolver = new LoginUserHandlerMethodArgumentResolver
    loginUserHandlerMethodArgumentResolver.setSessionService(sessionService)
    when(user.isGuest).thenReturn(false)
  }

  @Test def supportsParameter_false {
    when(parameter.hasParameterAnnotation(classOf[LoginUser])).thenReturn(false)
    assertThat(loginUserHandlerMethodArgumentResolver.supportsParameter(parameter), is(false))
  }

  @Test def supportsParameter_true {
    when(parameter.hasParameterAnnotation(classOf[LoginUser])).thenReturn(true)
    assertThat(loginUserHandlerMethodArgumentResolver.supportsParameter(parameter), is(true))
  }

  @Test(expected = classOf[LoginRequiredException])
  @throws(classOf[Exception])
  def loginUserRequired_but_guest {
    when(parameter.getParameterAnnotation(classOf[LoginUser])).thenReturn(new FakeLoginUser(true))
    when(sessionService.getLoginUser).thenReturn(SocialUser.GUEST_USER)
    loginUserHandlerMethodArgumentResolver.resolveArgument(parameter, null, webRequest, null)
  }

  @Test
  @throws(classOf[Exception])
  def loginUser_not_required {
    when(parameter.getParameterAnnotation(classOf[LoginUser])).thenReturn(new FakeLoginUser(false))
    when(sessionService.getLoginUser).thenReturn(SocialUser.GUEST_USER)
    val loginUser: SocialUser = loginUserHandlerMethodArgumentResolver.resolveArgument(parameter, null, webRequest, null).asInstanceOf[SocialUser]
    assertThat(loginUser, is(SocialUser.GUEST_USER))
  }

  @Test
  def loginUser_일반상황 {
    val user: SocialUser = new SocialUser
    when(parameter.getParameterAnnotation(classOf[LoginUser])).thenReturn(new FakeLoginUser(true))
    when(sessionService.getLoginUser).thenReturn(user)
    val loginUser: SocialUser = loginUserHandlerMethodArgumentResolver.resolveArgument(parameter, null, webRequest, null).asInstanceOf[SocialUser]
    assertThat(loginUser, is(user))
  }
}

